### ISSUE TITLE
why get nan while QAT

### ISSUE THREAD

**Issue by 2730gf** [microsoft/nni#4443]
**Describe the issue**:
When I quantify the resnet18 in QAT method, the loss will explode. This will happen no matter how many nodes are quantified.
The code is there:
`config_list = [{
        'quant_types': ['weight', 'input'],
        'quant_bits': {'weight': 8, 'input': 8,},
            'op_names': op_dict['conv2d']
        }, {
            'quant_types': ['output'],
            'quant_bits': {'output': 8, },
            'op_names': op_dict['relu']
        }, {
            'quant_types': ['output', 'weight', 'input'],
            'quant_bits': {'output': 8, 'weight': 8, 'input': 8},
            'op_names': op_dict['fc'],
        }]
dummy_input = torch.randn(3,3,32,32).to(device)
quantizer = QAT_Quantizer(model, config_list, optimizer, dummy_input=dummy_input)
quantizer.compress()`

**Environment**:
- NNI version:2.5
- Training service (local|remote|pai|aml|etc):
- Client OS:
- Server OS (for remote mode only):
- Python version:
- PyTorch/TensorFlow version:
- Is conda/virtualenv/venv used?:
- Is running in Docker?:

**Configuration**:
 - Experiment config (remember to remove secrets!):
 - Search space:

**Log message**:
 - nnimanager.log:
 - dispatcher.log:
 - nnictl stdout and stderr:
 

**How to reproduce it?**:

**Comment by J-shang** [comment]
Hello @2730gf , have you trained the model after `quantizer.compress()`, do you mean the test loss is nan? QAT is a quantization-aware training process, here is an example for using QAT https://github.com/microsoft/nni/blob/master/examples/model_compress/quantization/QAT_torch_quantizer.py

The training after `quantizer.compress()` is necessary. If you have already trained the model after `compress()`, but the train/test loss is also nan, please tell us.

**Comment by 2730gf** [comment]
Hello, I have used `quantizer.compress()`. And the difinition of resnet18 if from https://github.com/microsoft/nni/blob/master/examples/model_compress/models/cifar10/resnet.py.
After importing a trained model weight. The test loss is nan. Later in the training process, loss gradually becomes nan.
By the way, is also a bug in https://github.com/microsoft/nni/blob/12b5dbe2938847a2b1e6e8bcb9f972237af433b6/examples/model_compress/end2end_compression.py#L201. When converting to TensorRT engine, both "weight" and "input" are required.

**Comment by 2730gf** [comment]
> Hello, I have used `quantizer.compress()`. And the difinition of resnet18 if from https://github.com/microsoft/nni/blob/master/examples/model_compress/models/cifar10/resnet.py. After importing a trained model weight. The test loss is nan. Later in the training process, loss gradually becomes nan. By the way, is also a bug in
> 
> https://github.com/microsoft/nni/blob/12b5dbe2938847a2b1e6e8bcb9f972237af433b6/examples/model_compress/end2end_compression.py#L201
> 
> . When converting to TensorRT engine, both "weight" and "input" are required.

In the process of training, loss will become nan after one or two epochs.

**Comment by 2730gf** [comment]
> Hello, I have used `quantizer.compress()`. And the difinition of resnet18 if from https://github.com/microsoft/nni/blob/master/examples/model_compress/models/cifar10/resnet.py. After importing a trained model weight. The test loss is nan. Later in the training process, loss gradually becomes nan. By the way, is also a bug in
> 
> https://github.com/microsoft/nni/blob/12b5dbe2938847a2b1e6e8bcb9f972237af433b6/examples/model_compress/end2end_compression.py#L201
> 
> . When converting to TensorRT engine, both "weight" and "input" are required.

The code is here. https://github.com/2730gf/nni-validation/blob/main/nni_qat_practice.py
s there anything else I need to provide?

**Comment by J-shang** [comment]
> Hello, I have used `quantizer.compress()`. And the difinition of resnet18 if from https://github.com/microsoft/nni/blob/master/examples/model_compress/models/cifar10/resnet.py. After importing a trained model weight. The test loss is nan. Later in the training process, loss gradually becomes nan. By the way, is also a bug in
> 
> https://github.com/microsoft/nni/blob/12b5dbe2938847a2b1e6e8bcb9f972237af433b6/examples/model_compress/end2end_compression.py#L201
> 
> . When converting to TensorRT engine, both "weight" and "input" are required.

Thank you, yes, this is a bug, I will fix it.

And I tried your code, it seems to work well, the only difference is I change the config_list to
```
        config_list = [{
            'quant_types': ['weight', 'input'],
            'quant_bits': {'weight': 8, 'input': 8,},
            'op_types': ['Conv2d']
        }, {
            'quant_types': ['output', 'weight', 'input'],
            'quant_bits': {'output': 8, 'weight': 8, 'input': 8},
            'op_types': ['Linear']
        }]
```
because I don't know `get_op_names()` and no `ReLu` module in ResNet18. So I think maybe this is where cause problem.

**Comment by 2730gf** [comment]
I did it again as you did, and the result was the same, very weird, let me check again.

**Comment by 2730gf** [comment]
Could it be a problem with model weights? I pushed the model weight again, and now you can directly reproduce my operation with `python nni_qat_practice.py` .

**Comment by J-shang** [comment]
> Could it be a problem with model weights? I pushed the model weight again, and now you can directly reproduce my operation with `python nni_qat_practice.py` .

I tried with your pre-trained model, because you use a pre-trained model, better set `quant_start_step` and don't quantize the last layer right now.
```
        config_list = [{
            'quant_types': ['weight', 'input'],
            'quant_bits': {'weight': 8, 'input': 8,},
            'quant_start_step': 100,
            'op_types': ['Conv2d']}]
```

Because the quantizer will patch the optimizer, so I don't use a lr_scheduler, but maybe it is not the reason.

And make sure to load the model state dict before initializing the optimizer.

**Comment by scarlett2018** [comment]
> > Could it be a problem with model weights? I pushed the model weight again, and now you can directly reproduce my operation with `python nni_qat_practice.py` .
> 
> I tried with your model, because you use a pre-trained model, better set `quant_start_step` and don't quantize the last layer right now.
> 
> ```
>         config_list = [{
>             'quant_types': ['weight', 'input'],
>             'quant_bits': {'weight': 8, 'input': 8,},
>             'quant_start_step': 100,
>             'op_types': ['Conv2d']}]
> ```
> 
> Because the quantizer will patch the optimizer, so I don't use a lr_scheduler, but maybe it is not the reason.
> 
> And make sure to load the model state dict before initializing the optimizer.
> 
> After 10 epochs of training with `lr=0.001`, the final acc is 90.55%, and I think it can do better with a better hyperparameter, you can have a try.

@2730gf - 
Please press F12 to open browser console and check the error message.

@yzj19870824 - are you still bother by the problem? any feedback about the above suggestion. 
We are closing out of date issues, let us know if this is still a problem on your side, thanks.

**Comment by scarlett2018** [comment]
Closing the issue as overdue as it has no responds for a while, feel free to reopen if you are still seeing the problem on your side.
